### PR TITLE
i#1312 AVX-512 support: AVX-512 compiled client handling, get_xmm_vals() support.

### PR DESCRIPTION
**Pull request by hgreving2304**
Compile time detection has been added for whether a client was compiled with AVX-512. In this
case, and if DynamoRIO is not deployed via "earliest" inject, we initialize the lazy AVX-512 code 
detection to true, in order to prevent the client from clobbering potential AVX-512 application state.

Adds two new tests to check for the initialized state of AVX-512 lazy code detection. 

Provides support for zmm context in get_xmm_vals().

Provides support to dump zmm SIMD mcontext state. No tests for this have been added as this
is used for log files.

Issue: #1312

### CODE REVIEW

File: core/arch/arch.c [hunk 1]
```diff
@@ -3660,7 +3660,12 @@ dump_mcontext(priv_mcontext_t *context, file_t f, bool dump_xml)
     if (preserve_xmm_caller_saved()) {
```
Comment (derekbruening) [inline core/arch/arch.c]: > Provides support to dump zmm SIMD mcontext state. No tests has been added as this is used

grammar: s/has/have/

> No tests has been added as DynamoRIO itself does not clobber AVX-512 state. We explicitly

grammar: s/has/have/

**Comment by hgreving2304** [comment]
Holding off with committing this, the get_xmm_vals part needs more thought.

File: core/arch/arch.c (hunk 1 above)
Comment (derekbruening) [inline core/arch/arch.c]: > No tests has been added as DynamoRIO itself does not clobber AVX-512 state. We explicitly currently enforce that DynamoRIO is compiled without AVX-512 for portability and unintended frequency scaling.

Wait: what about the client?  DR invokes client events at various points.

File: core/arch/x86_code.c [hunk 2]
```diff
@@ -53,7 +53,9 @@ get_xmm_vals(priv_mcontext_t *mc)
 #ifdef X86
     if (preserve_xmm_caller_saved()) {
         ASSERT(proc_has_feature(FEATURE_SSE));
-        if (YMM_ENABLED())
+        if (ZMM_ENABLED())
```
Comment (derekbruening) [inline core/arch/x86_code.c]: It seems this needs to check the lazy cxt switching flag.  It is used on the initial thread for early injection and in other places where the lazy switch should apply.

Comment (hgreving2304) [reply core/arch/x86_code.c]: Done (offline). Compile time client AVX-512 code detection has been added, and the lazy switch is not initialized for the attach case if the client was AVX-512.

File: core/arch/arch.c (hunk 1 above)
Comment (hgreving2304) [reply core/arch/arch.c]: See above.

**Comment by hgreving2304** [comment]
run arm tests

**Comment by hgreving2304** [comment]
run arm tests

**Comment by hgreving2304** [comment]
The ARM error is unclear and seems unrelated.

**Comment by hgreving2304** [comment]
> The ARM error is unclear and seems unrelated.

Fixed.

File: core/dynamo.c [hunk 3]
```diff
@@ -383,7 +383,7 @@ get_dr_stats(void)
  * returns zero on success, non-zero on failure
  */
 DYNAMORIO_EXPORT int
-dynamorio_app_init(void)
+dynamorio_app_init(bool attach_case)
```
Comment (Carrotman42) [inline core/dynamo.c]: I think we need some docs about how to set this parameter correctly.

Are we OK with changing the public API like this? We probably at least want to make a change notice to the release doc.

Comment (hgreving2304) [reply core/dynamo.c]: It wouldn't be ok. But wait - this is not DR_API, this is DYNAMORIO_EXPORT. This shouldn't be public, is it?

Comment (derekbruening) [reply core/dynamo.c]: If this just indicates whether we're attaching, we already have dr_api_entry.

Comment (hgreving2304) [reply core/dynamo.c]: Ok! Then consider this reverted. It really just needs to know whether this is the attach case, right.

Comment (Carrotman42) [reply core/dynamo.c]: Oh yes, sorry! I confused tihs with "dr_app_setup" (which I was confused about, and thought was "dr_app_init").

Comment (hgreving2304) [reply core/dynamo.c]: @Carrotman42 , still a good call. I didn't like this flag at all.

Comment (derekbruening) [reply core/dynamo.c]: Don't you really need to know whether *any* app code has already executed?  If so, that happens with more than just attaching.  Non-earliest-inject will go and run system code.  It's possible that code will run AVX-512 instructions.

Comment (hgreving2304) [reply core/dynamo.c]: - see offline.

**Comment by hgreving2304** [comment]
This patch is on hold until we have discussed the deployment methods and how to distinguish them internally - offline.

File: core/dynamo.c (hunk 3 above)
Comment (hgreving2304) [reply core/dynamo.c]: Replaced with earliest inject flags, done.

**Comment by hgreving2304** [comment]
> This patch is on hold until we have discussed the deployment methods and how to distinguish them internally - offline.

PTAL.

File: core/lib/dr_api.h [hunk 4]
```diff
@@ -123,6 +123,20 @@ DR_EXPORT LINK_ONCE int _USES_DR_VERSION_ = ${VERSION_NUMBER_INTEGER};
 /* clang-format on */
 #endif
 
+/* AVX-512 client code detection. Compiling the client with AVX-512 will cause
+ * DynamoRIO to assume that AVX-512 code is in use when deploying DynamoRIO after
+ * the application has started per dr_app_setup().
+ */
+#ifndef DYNAMORIO_STANDALONE
+#    ifdef __AVX512F__
+DR_EXPORT LINK_ONCE bool _CLIENT_AVX512_CODE_IN_USE_ = true;
```
Comment (derekbruening) [inline core/lib/dr_api.h]: Use DR in the name.

File: core/dynamo.c [hunk 5]
```diff
@@ -668,6 +668,19 @@ dynamorio_app_init(void)
          */
         dr_app_started = create_broadcast_event();
 #ifdef CLIENT_INTERFACE
+#    ifdef X86
+#        ifdef WINDOWS
+        if (!dr_earliest_injected) {
+#        else
+        if (!dynamo_options.early_inject) {
+#        endif
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/dynamo.c]: Done.

File: core/dynamo.c [hunk 6]
```diff
@@ -668,6 +668,19 @@ dynamorio_app_init(void)
          */
         dr_app_started = create_broadcast_event();
 #ifdef CLIENT_INTERFACE
+#    ifdef X86
+#        ifdef WINDOWS
+        if (!dr_earliest_injected) {
+#        else
+        if (!dynamo_options.early_inject) {
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/dynamo.c]: Done with above.

File: core/lib/dr_api.h [hunk 7]
```diff
@@ -123,6 +123,20 @@ DR_EXPORT LINK_ONCE int _USES_DR_VERSION_ = ${VERSION_NUMBER_INTEGER};
 /* clang-format on */
 #endif
 
+/* AVX-512 client code detection. Compiling the client with AVX-512 will cause
+ * DynamoRIO to assume that AVX-512 code is in use when deploying DynamoRIO after
+ * the application has started per dr_app_setup().
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/lib/dr_api.h]: Done.

File: core/lib/instrument.c [hunk 8]
```diff
@@ -112,6 +112,9 @@ do_file_write(file_t f, const char *fmt, va_list ap);
  */
 DR_API const char *unique_build_number = STRINGIFY(UNIQUE_BUILD_NUMBER);
 
+/* The flag this name is referring to is described in arch.h. */
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/lib/instrument.c]: Done.

File: core/lib/dr_api.h (hunk 4 above)
Comment (hgreving2304) [reply core/lib/dr_api.h]: Done.

File: suite/tests/client-interface/avx512lazy.dll.c [hunk 9]
```diff
@@ -59,16 +68,23 @@ bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace, bool trans
...
+                if (seen_before) {
                     CHECK(dr_mcontext_zmm_fields_valid(),
                           "Error: dr_mcontext_zmm_fields_valid() should return true.");
                     dr_fprintf(STDERR, "After\n");
                 } else {
+                    /* The *-initial version of this test runs the client compiled with
+                     * AVX-512. Even in this case, the initial value of
+                     * dr_mcontext_zmm_fields_valid() is expected to be false. The only
+                     * time it should be true before application AVX-512 code has actually
+                     * been seen is the "attach' case. This is tested by another test
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/avx512lazy.dll.c]: Fixed.

File: core/dynamo.c [hunk 10]
```diff
@@ -668,6 +668,19 @@ dynamorio_app_init(void)
...
+#        endif
+            /* A client that had been compiled with AVX-512 may clobber an application's
+             * state. AVX-512 context switching will not be lazy in this case.
+             */
+            if (d_r_is_client_avx512_code_in_use())
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/dynamo.c]: Done.

File: suite/tests/client-interface/avx512lazy.dll.c [hunk 11]
```diff
@@ -59,16 +68,23 @@ bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace, bool trans
...
+                     * case.
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/client-interface/avx512lazy.dll.c]: Done.

File: core/lib/instrument.c [hunk 12]
```diff
@@ -576,6 +579,14 @@ add_client_lib(const char *path, const char *id_str, const char *options)
             SYSLOG(SYSLOG_ERROR, CLIENT_VERSION_INCOMPATIBLE, 2, get_application_name(),
                    get_application_pid());
         } else {
+#    ifdef X86
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/lib/instrument.c]: Done.

**Comment by hgreving2304** [comment]
PTAL.

**Pull request merged**